File: trusera-sdk-js/src/events.ts
/**
 * Event types tracked by the Trusera SDK.
 * These align with the core observability primitives for AI agent monitoring.
 */
export enum EventType {
  /** Tool or function call by the agent */
  TOOL_CALL = "tool_call",
  /** LLM inference invocation */
  LLM_INVOKE = "llm_invoke",
  /** Data access operation (read/write) */
  DATA_ACCESS = "data_access",
  /** Outbound HTTP API call */
  API_CALL = "api_call",
  /** File write operation */
  FILE_WRITE = "file_write",
  /** Decision point or chain step */
  DECISION = "decision",
}

/**
 * Core event structure for all Trusera tracking.
 * Events are immutable once created and queued for batch transmission.
 */
export interface Event {
  /** Unique event identifier (UUIDv4) */
  id: string;
  /** Event type discriminator */
  type: EventType;
  /** Human-readable event name (e.g., "openai.chat.completions", "github.api.repos") */
  name: string;
  /** Event-specific structured data */
  payload: Record<string, unknown>;
  /** Additional context (agent_id, session_id, tags, etc.) */
  metadata: Record<string, unknown>;
  /** ISO 8601 timestamp */
  timestamp: string;
}

/**
 * Creates a well-formed Event with automatic ID and timestamp generation.
 *
 * @param type - Event type from EventType enum
 * @param name - Descriptive event name (use dotted notation: "service.resource.action")
 * @param payload - Event-specific data (request, response, errors, etc.)
 * @param metadata - Optional metadata (merged with default agent context)
 * @returns Fully populated Event ready for tracking
 *
 * @example
 * ```typescript
 * const event = createEvent(
 *   EventType.API_CALL,
 *   "openai.chat.completions",
 *   { model: "gpt-4", tokens: 150 },
 *   { session_id: "abc-123" }
 * );
 * client.track(event);
 * ```
 */
export function createEvent(
  type: EventType,
  name: string,
  payload: Record<string, unknown> = {},
  metadata: Record<string, unknown> = {}
): Event {
  return {
    id: crypto.randomUUID(),
    type,
    name,
    payload: { ...payload },
    metadata: { ...metadata },
    timestamp: new Date().toISOString(),
  };
}

/**
 * Type guard to validate if an object is a valid Event.
 * Useful for runtime validation before transmission.
 */
export function isValidEvent(obj: unknown): obj is Event {
  if (typeof obj !== "object" || obj === null) return false;
  const e = obj as Partial<Event>;
  return (
    typeof e.id === "string" &&
    typeof e.type === "string" &&
    Object.values(EventType).includes(e.type as EventType) &&
    typeof e.name === "string" &&
    typeof e.payload === "object" &&
    e.payload !== null &&
    typeof e.metadata === "object" &&
    e.metadata !== null &&
    typeof e.timestamp === "string"
  );
}

File: trusera-agent-sdk/trusera_sdk/events.py
"""Event types and data structures for Trusera SDK."""

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class EventType(str, Enum):
    """Types of events that can be tracked by Trusera."""

    TOOL_CALL = "tool_call"
    LLM_INVOKE = "llm_invoke"
    DATA_ACCESS = "data_access"
    API_CALL = "api_call"
    FILE_WRITE = "file_write"
    DECISION = "decision"


@dataclass
class Event:
    """
    Represents a single event in an AI agent's execution.

    Attributes:
        type: The type of event (tool call, LLM invocation, etc.)
        name: Human-readable name of the event
        payload: Event-specific data (inputs, outputs, etc.)
        metadata: Additional context (duration, model, etc.)
        id: Unique identifier for the event
        timestamp: ISO 8601 timestamp of when the event occurred
    """

    type: EventType
    name: str
    payload: dict[str, Any] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())

    def to_dict(self) -> dict[str, Any]:
        """Convert the event to a dictionary for JSON serialization."""
        return {
            "id": self.id,
            "type": self.type.value,
            "name": self.name,
            "payload": self.payload,
            "metadata": self.metadata,
            "timestamp": self.timestamp,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Event":
        """Create an Event from a dictionary."""
        return cls(
            type=EventType(data["type"]),
            name=data["name"],
            payload=data.get("payload", {}),
            metadata=data.get("metadata", {}),
            id=data.get("id", str(uuid.uuid4())),
            timestamp=data.get("timestamp", datetime.now(timezone.utc).isoformat()),
        )
